### PR TITLE
Extend Cilium API to provide cluster status.

### PR DESCRIPTION
**Pull request by joestringer**
```release-note
`cilium status` now lists all of the cilium instances that it knows about.
```

The goal of this PR is to extend the Cilium API for getting health so that it provides a list of other cilium nodes that are known to this cilium node. This splits out the Cilium API parts from PR #2052, which will subsequently make use of this API and introduce a new daemon (and its own API) for connectivity debugging. Sending this separately for closer review of these API changes, and because this set of patches could be useful by themselves and are in a more ready state than the connectivity patches currently. This series also takes an initial stab at #2083, which was useful for testing & debugging the series.

The API changes are that under `GET "/healthz"`, the `StatusResponse` now has a new `ClusterStatus` which contains `self`, a name of the local node, and an array of `NodeElement`s. The `NodeElement` consists of a node name, primary address, and optional array of secondary addresses. The primary address specifies IPv4 and/or IPv6 addresses, while the secondary addresses each specify a single address which may be IPv4 or IPv6. This series also extends the `NodeAddressingElement` to add an optional address type, which determines whether the address is a `HostName` type, `ExternalIP`, or `InternalIP`. No attempt was made to extend existing usage of `NodeAddressingElement` to specify this new address type field.

After this commit, there is a new section at the end of the output of `cilium status`, "Known cluster nodes" as seen below:

```
$ cilium status
KVStore:            Ok         Consul: 172.17.0.2:8300
ContainerRuntime:   Ok
Kubernetes:         Disabled
Kubernetes APIs:    [""]
Cilium:             Ok   OK
NodeMonitor:        Disabled
Allocated IPv4 addresses:
 10.11.28.238
 10.11.247.232
Allocated IPv6 addresses:
 f00d::a0f:0:0:8ad6
Known cluster nodes:
 cilium-master (localhost):
  Primary Address:   10.0.2.15
   Type:             InternalIP
  AllocRange:        10.11.0.0/16
```

### CODE REVIEW

File: cilium/cmd/status.go [hunk 1]
```diff
@@ -42,6 +44,15 @@ func statusDaemon(cmd *cobra.Command, args []string) {
 	if resp, err := client.Daemon.GetHealthz(nil); err != nil {
 		fmt.Fprintf(os.Stderr, "%s\n", pkg.Hint(err))
 		os.Exit(1)
+	} else if viper.GetBool("json") {
+		result, err := json.MarshalIndent(resp.Payload, "", "  ")
+		if err != nil {
+			Fatalf("Cannot fetch status: %s", err.Error())
```
Comment (tgraf) [inline cilium/cmd/status.go]: Can you make this error message specific to Marshal failing?

**Review by tgraf — DISMISSED** [review]
Looks great overall, just a small nitpick around error messagae.

**Review by tgraf — APPROVED** [review]

**Review by tgraf — CHANGES_REQUESTED** [review]
Needs rebase

**Review by ianvernon — APPROVED** [review]

**Comment by joestringer** [comment]
Superseded by #2052.

**Pull request closed without merging**